src/jsoncanvas.rs

```diff
@@ -61,38 +76,33 @@ where
     D: Deserializer<'de>,
 {
     let vec: Vec<Edge> = Vec::deserialize(deserializer)?;
-    let map: HashMap<String, Edge> = vec.into_iter().map(|node| (node.id.clone(), node)).collect();
+    let map: HashMap<String, Edge> = vec
+        .into_iter()
+        .map(|node| (node.id.clone(), node))
+        .collect();
     Ok(map)
 }
 
-
 impl JsonCanvas {
-    pub fn new() -> JsonCanvas {
-        JsonCanvas {
-            nodes: HashMap::new(),
-            edges: HashMap::new()
-        }
-    }
-
     pub fn add_node(&mut self, node: Node) -> Result<(), JsonCanvasError> {
         if self.nodes.contains_key(node.id()) {
-            return Err(JsonCanvasError::AlreadyExists);
+            return Err(JsonCanvasError::NodeExists(node.id().to_string()));
         }
-        self.nodes.insert(node.id().clone(), node);
+        self.nodes.insert(node.id().to_string(), node);
         Ok(())
     }
 
-    pub fn add_edge(&mut self, edge: Edge) -> Result<(), JsonCanvasError>{
+    pub fn add_edge(&mut self, edge: Edge) -> Result<(), JsonCanvasError> {
         if self.edges.contains_key(&edge.id) {
-            return Err(JsonCanvasError::AlreadyExists);
+            return Err(JsonCanvasError::EdgeExists(edge.id().to_string()));
         }
 
         if !self.nodes.contains_key(&edge.from_node) {
-            return Err(JsonCanvasError::NodeNotExists);
+            return Err(JsonCanvasError::NodeNotExists(edge.from_node));
         }
 
         if !self.nodes.contains_key(&edge.to_node) {
-            return Err(JsonCanvasError::NodeNotExists);
+            return Err(JsonCanvasError::NodeNotExists(edge.from_node));
```

Comment:
There is a typo here, should be edge.to_node

Reply:
Fixed